components/experimental/src/duration/format.rs

```diff
@@ -0,0 +1,454 @@
+            self.fmt.options.hour == FieldStyle::Numeric
+                || self.fmt.options.hour == FieldStyle::TwoDigit
+        );
+
+        let mut fd = FixedDecimal::from(hour_val);
+        match self.fmt.options.hour {
+            FieldStyle::TwoDigit => fd.pad_start(2),
+            FieldStyle::Numeric => fd.pad_start(
+                self.get_numeric_unit_padding(
+                    Unit::Hour,
+                    hours_formatted,
+                    minute_formatted,
+                    second_formatted,
+                )
+                .into(),
+            ),
+            _ => unreachable!(),
+        }
+        if *sign_displayed {
+            fd.apply_sign_display(SignDisplay::Never);
+        }
+        let ffd = self.fmt.fdf.format(&fd);
+        sink.with_part(parts::HOUR, |e| ffd.write_to_parts(e))?;
+
+        Ok(())
+    }
+
+    fn format_numeric_minutes<V: PartsWrite + ?Sized>(
+        &self,
+        minute_val: u64,
+        sign_displayed: &mut bool,
+        hours_formatted: bool,
+        minute_formatted: bool,
+        second_formatted: bool,
+        sink: &mut V,
+    ) -> core::fmt::Result {
+        if hours_formatted {
+            sink.with_part(parts::LITERAL, |e| {
+                e.write_str(self.fmt.digital.get().separator.as_ref())
+            })?;
+        }
+
+        debug_assert!(
+            self.fmt.options.minute == FieldStyle::Numeric
+                || self.fmt.options.minute == FieldStyle::TwoDigit
+        );
+
+        let mut fd = FixedDecimal::from(minute_val);
+        if let FieldStyle::TwoDigit = self.fmt.options.hour {
+            fd.pad_start(2);
+        } else {
+            // Divergence from ECMAScript TC39 proposal:
+            // We pad the number based on cldr-json padding instead of not applying any padding.
+            fd.pad_start(
+                self.get_numeric_unit_padding(
+                    Unit::Minute,
+                    hours_formatted,
+                    minute_formatted,
+                    second_formatted,
+                )
+                .into(),
+            );
+        }
+        if !*sign_displayed {
+            fd.apply_sign_display(SignDisplay::Never);
+        }
+        let ffd = self.fmt.fdf.format(&fd);
+        sink.with_part(parts::MINUTE, |e| ffd.write_to_parts(e))?;
+
+        Ok(())
+    }
+
+    fn seconds_with_fractional_digits(&self) -> FixedDecimal {
+        // Convert to u128 to avoid overflow
+        let mut result: u128 = self.duration.seconds.into();
+        result *= u128::pow(10, 9);
+
+        for (style, value, exp) in [
+            (self.fmt.options.millisecond, self.duration.milliseconds, 6),
+            (self.fmt.options.microsecond, self.duration.microseconds, 3),
+            (self.fmt.options.nanosecond, self.duration.nanoseconds, 0),
+        ] {
+            if style == FieldStyle::Fractional {
+                result += u128::from(value) * u128::pow(10, exp);
+            }
```

Comment:
Do we disallow values that are >= 1000, or do we rewrite this code handle overlapping magnitudes differently?